src/crimson/os/alienstore/alien_store.cc

```diff
@@ -0,0 +1,380 @@
+      } else {
+        return get_attr_errorator::make_ready_future<ceph::bufferptr>(
+          std::move(value));
+      }
+    });
+  });
+}
+
+AlienStore::get_attrs_ertr::future<AlienStore::attrs_t>
+AlienStore::get_attrs(CollectionRef ch,
+                      const ghobject_t& oid)
+{
+  logger().debug("{}", __func__);
+  return seastar::do_with(attrs_t{}, [=] (auto &aset) {
+    return tp->submit([=, &aset] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->getattrs(c->collection, oid,
+		             reinterpret_cast<map<string,bufferptr>&>(aset));
+    }).then([&aset] (int r) -> get_attrs_ertr::future<attrs_t> {
+      if (r == -ENOENT) {
+        return crimson::ct_error::enoent::make();;
+      } else {
+        return get_attrs_ertr::make_ready_future<attrs_t>(std::move(aset));
+      }
+    });
+  });
+}
+
+seastar::future<AlienStore::omap_values_t>
+AlienStore::omap_get_values(CollectionRef ch,
+                       const ghobject_t& oid,
+                       const set<string>& keys)
+{
+  logger().debug("{}", __func__);
+  return seastar::do_with(omap_values_t{}, [=] (auto &values) {
+    return tp->submit([=, &values] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->omap_get_values(c->collection, oid, keys,
+		                    reinterpret_cast<map<string, bufferlist>*>(&values));
+    }).then([&values] (int) {
+      return seastar::make_ready_future<omap_values_t>(std::move(values));
+    });
+  });
+}
+
+seastar::future<bool, AlienStore::omap_values_t>
+AlienStore::omap_get_values(CollectionRef ch,
+                            const ghobject_t &oid,
+                            const std::optional<string> &start)
+{
+  logger().debug("{} with_start", __func__);
+  return seastar::do_with(omap_values_t{}, [=] (auto &values) {
+    return tp->submit([=, &values] {
+      auto c = static_cast<AlienCollection*>(ch.get());
+      return store->omap_get_values(c->collection, oid, start,
+		                    reinterpret_cast<map<string, bufferlist>*>(&values));
+    }).then([&values] (int r) {
+      return seastar::make_ready_future<bool, omap_values_t>(true, std::move(values));
+    });
+  });
+}
+
+seastar::future<> AlienStore::do_transaction(CollectionRef ch,
+                                             ceph::os::Transaction&& txn)
+{
+  logger().debug("{}", __func__);
+  auto callback =
+    std::make_unique<OnCommit>(seastar::engine().cpu_id(), txn);
+  return seastar::do_with(std::move(txn), std::move(callback),
+      [this, ch] (ceph::os::Transaction &txn, auto &callback) {
+    return seastar::with_gate(transaction_gate, [this, ch, &txn, &callback] {
+      return tp_mutex.lock().then ([this, ch, &txn, &callback] {
+        return tp->submit([=, &txn, &callback] {
+          txn.register_on_commit(callback.get());
+          auto c = static_cast<AlienCollection*>(ch.get());
+          return store->queue_transaction(c->collection, std::move(txn));
```

Comment:
Here, you're directly queuing the Transaction with it's nested Contexts.  Why not pull out those contexts and wrap them in something that bluestore can safely use and delete?

Reply:
I need a complete context to inform crimson bluestore finished the process then the continuation can go on. so the context should keep alive and can't be deleted in bluestore. That is the simply way I can get.  could you please give me some suggestion what mechanism we can use here?

Reply:
I missed where you are already wrapping them in the OnCommit callback.

I think something like https://github.com/athanatos/ceph/tree/sjust/ceph_seastar_alien_blue_store might solve the problem -- it allocates the context in bluestore's thread so that bluestore will only release the context itself -- everything else gets released from the crimson reactor.

What vstart.sh command and config file are you using to test?

Reply:
the following is the cmmmand I used.  config fie is vstart default 
MDS=0 MGR=1 OSD=1 MON=1 ../src/vstart.sh -n --without-dashboard --bluestore -X  --crimson
bin/ceph osd pool create rbd 128
bin/rbd create fio_test --size 10G --image-format=2 --rbd_default_features=3
LD_LIBRARY_PATH=lib/ fio ../rbd_write.fio

Reply:
@athanatos I have committed your patch in this PR.